### ISSUE TITLE
AbstractInjectionDetector.checkTaintSink() modifies Set<TaintSink> while iterating over it

### ISSUE THREAD

**Issue by mkienenb** [find-sec-bugs/find-sec-bugs#177]
Looks like the Set<TaintSink> being iterated over in AbstractInjectionDetector.checkTaintSink() is being modified.   I can see that delayBugToReport() is one place that additional TaintSinks are added to such a Set.  It's unclear to me if there are other places where the set might be modified.

It's also unclear to me whether it is important for the code in the loop to be notified of changes to the TaintSink Set so I can't propose a fix.

```
 [findbugs]   Exception while analyzing com.xyz.utilities.business.FlexList.getDescendents(JLjava/lang/String;Ljava/lang/String;I)Lcom/xyz/utilities/business/FlexList;
 [findbugs]     java.util.ConcurrentModificationException
 [findbugs]       At java.util.HashMap$HashIterator.nextNode(HashMap.java:1429)
 [findbugs]       At java.util.HashMap$KeyIterator.next(HashMap.java:1453)
 [findbugs]       At com.h3xstream.findsecbugs.injection.AbstractInjectionDetector.checkTaintSink(AbstractInjectionDetector.java:125)
 [findbugs]       At com.h3xstream.findsecbugs.injection.AbstractInjectionDetector.analyzeLocation(AbstractInjectionDetector.java:83)
 [findbugs]       At com.h3xstream.findsecbugs.injection.AbstractTaintDetector.analyzeMethod(AbstractTaintDetector.java:106)
 [findbugs]       At com.h3xstream.findsecbugs.injection.AbstractTaintDetector.visitClassContext(AbstractTaintDetector.java:75)
 [findbugs]       At edu.umd.cs.findbugs.DetectorToDetector2Adapter.visitClass(DetectorToDetector2Adapter.java:76)
 [findbugs]       At edu.umd.cs.findbugs.FindBugs2.analyzeApplication(FindBugs2.java:1089)
 [findbugs]       At edu.umd.cs.findbugs.FindBugs2.execute(FindBugs2.java:283)
 [findbugs]       At edu.umd.cs.findbugs.FindBugs.runMain(FindBugs.java:402)
 [findbugs]       At edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1200)

```

**Comment by mkienenb** [comment]
I can't recommend this as a fix for the reasons stated above, but here is a temporary workaround I am using:

diff --git a/plugin/src/main/java/com/h3xstream/findsecbugs/injection/AbstractInjectionDetector.java b/plugin/src/main/java/com/h3xstream/findsecbugs/injection/AbstractInjection
Detector.java
index c727d6d..d21c9a3 100644
--- a/plugin/src/main/java/com/h3xstream/findsecbugs/injection/AbstractInjectionDetector.java
+++ b/plugin/src/main/java/com/h3xstream/findsecbugs/injection/AbstractInjectionDetector.java
@@ -122,7 +122,12 @@ public abstract class AbstractInjectionDetector extends AbstractTaintDetector {
     
     private void checkTaintSink(ConstantPoolGen cpg, InvokeInstruction invoke, TaintFrame fact,
             SourceLineAnnotation sourceLine, String currentMethod) throws DataflowAnalysisException {
-        for (TaintSink sink : getSinks(cpg, invoke, fact)) {
+        Set<TaintSink> originalSet = getSinks(cpg, invoke, fact);
+        Set<TaintSink> copiedSet = null;
+        if (null != originalSet) {
+            copiedSet = new HashSet<TaintSink>(originalSet);
+        }
+        for (TaintSink sink : copiedSet) {
             Taint sinkTaint = sink.getTaint();
             Set<Integer> taintParameters = sinkTaint.getParameters();
             Taint finalTaint = Taint.valueOf(sinkTaint.getNonParametricState());

**Comment by formanek** [comment]
@mkienenb Thank you for reporting. I rewritten this part recently (it is waiting for merge) and the new code shouldn't modify collections during iteration. You can test [my fork](https://github.com/formanek/find-sec-bugs).

**Comment by mkienenb** [comment]
Verified that this was fixed by pull request #173
